src/NLog/Layouts/Typed/Layout.cs

```diff
@@ -0,0 +1,315 @@
+        {
+            return new Layout<T>(layout);
+        }
+
+        #endregion
+
+        /// <inheritdoc />
+        protected override string GetFormattedMessage(LogEventInfo logEvent)
+        {
+            if (_fixedValue)
+            {
+                if (_value == null)
+                {
+                    return null;
+                }
+
+                var text = ValueToString(_value, LoggingConfiguration.DefaultCultureInfo);
+                if (text != null)
+                {
+                    return text;
+                }
+            }
+
+            return _layout.Render(logEvent);
+        }
+
+        /// <inheritdoc cref="IRawValue" />
+        internal override bool TryGetRawValue(LogEventInfo logEvent, out object rawValue)
+        {
+            if (_fixedValue)
+            {
+                rawValue = _value;
+                return true;
+            }
+
+            if (_layout == null)
+            {
+                rawValue = null;
+                return true;
+            }
+
+            if (_layout.TryGetRawValue(logEvent, out var raw))
+            {
+                var success = TryConvertRawToValue(raw, out var i);
+                rawValue = i;
+                return success;
+            }
+
+            rawValue = null;
+            return false;
+        }
+
+        /// <summary>
+        /// Render to value
+        /// </summary>
+        /// <returns></returns>
+        T IRenderable<T>.RenderToValue(LogEventInfo logEvent)
+        {
+            return RenderToValueInternal(logEvent, null);
+        }
+
+        /// <summary>
+        /// Render to value
+        /// </summary>
+        /// <param name="logEvent"></param>
+        /// <param name="reusableBuilder">if null, default layout render will be used</param>
+        /// <returns></returns>
+        internal T RenderToValueInternal(LogEventInfo logEvent, [CanBeNull] StringBuilder reusableBuilder)
+        {
+            if (_fixedValue)
+            {
+                return _value;
+            }
+
+            if (_layout == null)
+            {
+                return default(T);
+            }
+
+            if (_layout.TryGetRawValue(logEvent, out var raw))
+            {
+                if (TryConvertRawToValue(raw, out var value))
+                {
+                    return value;
+                }
+
+                InternalLogger.Warn("rawvalue isn't a {0} ", _typeNamed);
+            }
+
+            var text = reusableBuilder != null ? RenderAllocateBuilder(logEvent, reusableBuilder) : _layout.Render(logEvent);
+            if (TryParse(text, out var parsedValue))
+            {
+                return parsedValue;
+            }
+
+            InternalLogger.Warn("Parse {0} to {1} failed", text, _typeNamed);
+            return default(T);
+        }
+
+        private static bool TryConvertRawToValue(object raw, out T value)
+        {
+            if (raw == null)
```

Comment:
Not sure I like that null-value automatically translate into default integer/enum-value. Might be working for reference-class-types.

Reply:
well this is needed if T is just a int, we can;t assign `null` to it?

Reply:
Trying to say that it should fail when converting null into an integer (Unless nullable).

Reply:
fixed